bindata/assets/kube-apiserver/pod.yaml

```diff
@@ -19,16 +19,44 @@ spec:
       volumeMounts:
         - mountPath: /var/log/kube-apiserver
           name: audit-dir
-      command: ['/usr/bin/timeout', '105', '/bin/bash', '-ec'] # a bit more than 60s for graceful termination + 35s for minimum-termination-duration, 5s extra cri-o's graceful termination period
+      command: ['/usr/bin/timeout', '200', '/bin/bash', '-ec'] # a bit more than 60s for graceful termination + 35s for minimum-termination-duration, 5s extra cri-o's graceful termination period
       args:
       - |
-        echo -n "Fixing audit permissions."
+        echo -n "Fixing audit permissions ..."
         chmod 0700 /var/log/kube-apiserver && touch /var/log/kube-apiserver/audit.log && chmod 0600 /var/log/kube-apiserver/*
-        echo -n "Waiting for port :6443 and :6080 to be released."
-        while [ -n "$(ss -Htan '( sport = 6443 or sport = 6080 )')" ]; do
-          echo -n "."
-          sleep 1
-        done
+
+        LOCK=/var/log/kube-apiserver/.lock
+        echo -n "Acquiring exclusive lock ${LOCK} ..."
+
+        # Waiting for 130s max for old kube-apiserver's watch-termination process to exit and remove the lock.
+        # Two cases:
+        # 1. if kubelet does not start the old and new in parallel (i.e. works as expected), the flock will always succeed without any time.
+        # 2. if kubelet does overlap old and new pods for up to 130s, the flock will wait and immediate return when the old finishes.
+        #
+        # NOTE: We can increase 130s for a bigger expected overlap. But a higher value means less noise about the broken kubelet behaviour, i.e. we hide a bug.
+        # NOTE: Do not tweak these timings without considering the livenessProbe initialDelaySeconds
+        exec {LOCK_FD}>${LOCK} && flock --verbose -w 130 "${LOCK_FD}" || {
+          echo "$(date -Iseconds -u) kubelet did not terminate old kube-apiserver before new one" >> /var/log/kube-apiserver/lock.log
+          echo -n ": WARNING: kubelet did not terminate old kube-apiserver before new one."
+
+          # We failed to acquire exclusive lock, which means there is old kube-apiserver running in system.
+          # We don't want to fail here and we will continue, risking audit log corruption and termination log corruption.
+          # Since we utilize SO_REUSEPORT, we need to make sure the old kube-apiserver stopped listening.
```

Comment:
is not the opposite, since we use SO_REUSEPORT we can bind despite it is still listening?
or you want to mean that since we use SO_REUSEPORT we don't want to take over the traffic until the old has left?

Reply:
@aojea we never want to take over the traffic, we want to make sure the port is not used by other process. issue is, kernel takes time (~60s) to do this for us.

Reply:
ok, then the comment is a bit confusing to me,AFAIK:
1. if previous process bound with the SO_REUSEPORT, new process can bind to the port and traffic will be split
2. if previous process bound without the SO_REUSEPORT, it simply will fail to bind
3. before closing, the port has to go through the TIME_WAIT status, that is governed by `net.ipv4.tcp_fin_timeout` (60s in my host)

However, if the leaving apiserver is in time_wait, I think that you don't take over the traffic, just the new traffic is sent to your new socket only, because is the only one to accept() connections.